### ISSUE TITLE
Consider different pattern replacement syntax

### ISSUE THREAD

**Issue by Integralist** [Integralist/go-reverse-proxy#1]
Daniel Katz (@katzdm, colleague at BuzzFeed) suggested...

> It might be that the regex `<cap>`-capturing syntax is something standard that I haven't come across before, but I found it a bit awkward to read. An implementation of a similar matching system which I personally like is implemented in Google's Copybara library for moving source code between repositories (http://github.com/google/copybara). In particular, I'm a fan of their regex syntax used in the `core.replace` transformation.

```py
core.replace(
       before = "https://some_internal/url/${pkg}.html",
       after = "https://example.com/${pkg}.html",
       regex_groups = {
           "pkg": ".*",
       },
   )
```

> The `regex_groups` syntax has the advantage of allowing for the definition of arbitrarily many capture-groups, and uses a ubiquitous syntax (i.e. `${var-name}`) in the before/after fields, arguably making the before/after patterns a bit easier on the eyes.

**Comment by katzdm** [comment]
Hey Mark - Thanks for opening an issue for this! Here's an example of what I had in mind, to help to make the suggestion more concrete.

The following is inspired by a [current routing configuration](https://github.com/Integralist/go-reverse-proxy/blob/master/routing/configuration.go#L34-L42) in go-reverse-proxy, with more sophisticated regex patterns added to better demonstrate the differences:

```
Config{
	Path:     "/(?P<start>anything\d{1,3})/(?P<cap>foobar\d{1,3})$",
	Upstream: upstreams.HTTPBin,
	Override: Override{
		Header:     "X-BF-Testing",
		Match:      "integralist",
		ModifyPath: "/anything/newthing$cap",
	},
}
```

What do you think of a syntax like this, instead?

```
Config{
	Upstream: upstreams.HTTPBin,
	Path:     "/${anything}/${cap}$",
	RegexGroups: map[string]string{
		"anything": "anything\d{1,3}",
		"cap": "foobar\d{1,3}",
	},
	Override: Override{
		Header:     "X-BF-Testing",
		Match:      "integralist",
		ModifyPath: "/anything/newthing${foobar}",
	},
}
```

Note that this could be transformed into the standard PCRE syntax prior to feeding expressions into `Regexp.Expand()` by iterating over `RegexGroups`, and for each (K, V) replacing every `${K}` with `(?P<K>V)`.

The suggestion is intended to help improve the readability of `Path`-expressions in configs, but that might be a simple matter of taste - Even if you agree that the proposed syntax is prettier, I leave it to you to decide if it's worth the added complexity to implement :-)

**Comment by Integralist** [comment]
Hey @katzdm thanks again for opening this issue.

So to clarify, I do prefer the replacement syntax of `${foo}` over what I had used (`$foo`). I think the reason why I originally used `$foo` was because that was something that nginx utilised (although nginx also supports `${foo}` and I never noticed!?) and I just got used to seeing that form.

That said I've updated the code to use your suggestion: 

https://github.com/Integralist/go-reverse-proxy/commit/c3367a0b289b915cbbf41b1e90c94c5cf2c8456d

I decided _not_ to change the routing configuration design though. The reason being is that the routing format is just plain 'ole standard regular expression syntax and I personally prefer that.

I agree that the suggestion of using `${...}` in the routing segment does _look_ cleaner, but I'm not overly keen on moving too far away from the standard regex syntax, and also it would require an additional configuration block (i.e. `RegexGroups`) which feels like a little too much extra complexity for simply changing the routing format (which is something I'm not overly worried about).

But thanks again for opening the issue because it did make me realise the replacement syntax needed updating to what you suggested 👍
